### ISSUE TITLE
backup of file whose blocklist is in backup as file doesn't put it in dindex list folder, causing recreate to read dblock

### ISSUE THREAD

**Issue by ts678** [duplicati/duplicati#5038]
- [x] I have searched open and closed issues for duplicates.
- [x] I have searched the [forum](https://forum.duplicati.com) for related topics.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.7.2 Canary
- **Operating system**: Windows 10
- **Backend**: Local folder

## Description

Slight recreate delay due to pass 1 download of dblock file during DB recreate because no dindex file claims blocklist as a blocklist, and it also throws off checker12.py as an anomaly. I might have to turn down its sensitivity if this is considered proper-but-strange behavior...

[re-created dindex file lacked list folder, so DB recreate ran pass 1 of dblock downloads to get a blocklist #5035](https://github.com/duplicati/duplicati/issues/5035) might be similar, but was overall so complex that it was easier to write a separate test of hypothesis that Duplicati mishandles the case where a file gets backed up but its blocklist is already in the backup as an ordinary file. It's not putting the blocklist in the dblock. What about a dindex `list` entry?

## Steps to reproduce

1. backup two block file
1. extract the blocklist
1. setup for new backup
1. backup the blocklist
1. backup two block file

- **Actual result**:
dindex has no list folder
- **Expected result**:
dindex has list folder containing a copy of the blocklist, however that should be thought through. It might be safer to leave it as is...

## Debug log

Verbose live log:

```
Oct 1, 2023 2:10 PM: The operation Repair has completed
Oct 1, 2023 2:10 PM: Setting custom SQLite option 'cache_size=-200000'.
Oct 1, 2023 2:10 PM: Recreate completed, and consistency checks completed, marking database as complete
Oct 1, 2023 2:10 PM: Recreate completed, verifying the database consistency
Oct 1, 2023 2:10 PM: Pass 1 of 3, processing blocklist volume 1 of 1
Oct 1, 2023 2:10 PM: Backend event: Get - Completed: duplicati-bd6eaa316e96c4a2cb8beb5420bf87ca1.dblock.zip (721 bytes)
Oct 1, 2023 2:10 PM: Backend event: Get - Started: duplicati-bd6eaa316e96c4a2cb8beb5420bf87ca1.dblock.zip (721 bytes)
Oct 1, 2023 2:10 PM: Processing required 1 blocklist volumes
Oct 1, 2023 2:10 PM: Processing required 1 blocklist volumes: duplicati-bd6eaa316e96c4a2cb8beb5420bf87ca1.dblock.zip
Oct 1, 2023 2:10 PM: Processing indexlist volume 2 of 2
Oct 1, 2023 2:10 PM: Backend event: Get - Completed: duplicati-ideff5d5ef3d9404aa4f29a7249af48d7.dindex.zip (650 bytes)
Oct 1, 2023 2:10 PM: Backend event: Get - Started: duplicati-ideff5d5ef3d9404aa4f29a7249af48d7.dindex.zip (650 bytes)
Oct 1, 2023 2:10 PM: Processing indexlist volume 1 of 2
Oct 1, 2023 2:10 PM: Backend event: Get - Completed: duplicati-i531b6583adfb4f2aa541a06d59d47a1c.dindex.zip (609 bytes)
Oct 1, 2023 2:10 PM: Backend event: Get - Started: duplicati-i531b6583adfb4f2aa541a06d59d47a1c.dindex.zip (609 bytes)
Oct 1, 2023 2:10 PM: Filelists restored, downloading 2 index files
Oct 1, 2023 2:10 PM: Processing filelist volume 2 of 2
Oct 1, 2023 2:10 PM: Backend event: Get - Completed: duplicati-20231001T180403Z.dlist.zip (767 bytes)
Oct 1, 2023 2:10 PM: Processing filelist volume 1 of 2
Oct 1, 2023 2:10 PM: Backend event: Get - Started: duplicati-20231001T180403Z.dlist.zip (767 bytes)
Oct 1, 2023 2:10 PM: Backend event: Get - Completed: duplicati-20231001T180325Z.dlist.zip (640 bytes)
Oct 1, 2023 2:10 PM: Backend event: Get - Started: duplicati-20231001T180325Z.dlist.zip (640 bytes)
Oct 1, 2023 2:10 PM: Rebuild database started, downloading 2 filelists
Oct 1, 2023 2:10 PM: Backend event: List - Completed: (6 bytes)
Oct 1, 2023 2:10 PM: Backend event: List - Started: ()
Oct 1, 2023 2:10 PM: Setting custom SQLite option 'cache_size=-200000'.
Oct 1, 2023 2:10 PM: The operation Repair has started
```

```
checker12.py "C:\Duplicati\duplicati-2.0.7.2_canary_2023-05-25\RUN\test 2"
5 blocks seen in dindex files
0 blocklists seen in dindex files
1 blocks used by dlist files blocklists
Missing {'TE22wCHk2ZCMDmPgRmZrPOQY2tifA17OQOkzUAgwj8U='}
0 blocks used by dlist large files data
1 blocks used by dlist small files data
2 blocks used by dlist files metadata
2 blocks unused
1 large blocksets in dlist files
1 small blocksets in dlist files
small file blocksets that are also metadata blocksets: set()
small file blocksets that are also blocklists: set()
```

`TE22wCHk2ZCMDmPgRmZrPOQY2tifA17OQOkzUAgwj8U=` is the 64 byte extracted-and-backed-up (but not identified as blocklist) blocklist file.

**Comment by Jojo-1000** [comment]
I think the problem is that a normal block (say block A uploaded in a.dblock with a.dindex) will become a blocklist at a later stage (with b.dblock and b.dindex containing the linked blocks B and C, but not the original blocklist data A).

At the time a.dindex is created, A is a normal file block and should not be duplicated in the list folder.
When b.dindex is created, the corresponding b.dblock only contains B and C, but not A so it is not contained in the index.

Potential ways to fix this:
- modify a.dindex *after* it was originally created (upload new version and delete old one, with potential transaction issues).
- It might also be possible to create a a2.dindex that links to a.dblock and only contains A in the list folder, but I am not sure if multiple indexes per block are fully bug-free (should be supported if I recall correctly, but it usually does not happen).
- b.dindex could contain the blocklist A anyways, but it could cause problems because there is no way in the format to indicate in which dblock A actually is (that might also be fine). This would potentially be difficult if b.dblock is deleted or compacted, but A still exists in a.dblock.

In all of these cases, we would need to detect that  block A was not a blocklist before and now is, including that it *never has been* a blocklist in a previously deleted version.

Does this edge case occur often? I think having a raw blocklist in your backup data would be reasonably rare under normal circumstances.

Also this should not happen if blocklist is added after the original file. When both appear in the same version, it will probably depend on the processing order and whether they are put in the same dblock.

**Comment by ts678** [comment]
Excellent points. The 64 byte file (a pair of 32 byte SHA-256 hashes for my minimal case -- more could be used) is a single block. Promoting that to also be a blocklist while meeting all the I'm-not-sure-what-they-are assumptions in the code is the challenge. Simply by looking at actual dindex files with a tool like checker12.py finds that duplicate blocklists are normal, and there's even a

https://github.com/duplicati/duplicati/blob/f6440a8e706b3c30ac430205853b21d50796d6cd/Duplicati/Library/Main/Operation/Backup/SpillCollectorProcess.cs#L109-L112

Regardless, my backup is full of duplicate blocklists, but I haven't done an analysis to see exactly where they're scattered around. What I did yesterday was to hide those, and the duplicate blocks that are probably from not recycling out of DeletedBlock table. Also hidden are the minor infraction here, where the penalty is a dblock read. Any blocklist that's actually missing is still an error.

I have no idea how often this edge case happens. Probably rarely, but I don't know how one could survey all of the real backups. Primary goal IMO is to focus on finding and fixing the big corruptions that somehow happen, but so far this doesn't look so bad. Maybe what I'll do (having relaxed some of the noise from the checker script) is just put that backup back in service to run more.

So far the elusive goal of finding how to break a backup badly is unfortunately not met, but lesser breaks/issues are showing up.

> Also this should not happen if blocklist is added after the original file.

I'm not sure quite how to read that, but if you mean one backup with the two block file and its 64 bit blocklist works, maybe so, however I haven't tested. Feel free. I also need to look at the other issue some more, on creation of the new file and its blocklist.

**Comment by ts678** [comment]
I verified that putting steps 4 (the blocklist) and 5 (the file) together into one backup works. I made a one-byte edit to the 64 byte blocklist for an issue I noticed later that's keeping my production backup from going back into service. Recreate leaves the dblock from first backup as Uploaded. This is the one that it must read in order to read the blocklist that the second dindex file didn't list.

If I do the original test but using edited blocklist, the confusion of this issue doesn't happen, and all files are `Verified` as is usual. Arguably this could get a different issue, but it's so closely tied to the repro steps here that I'm going to start it just with this note.

I have no reason to suspect that the exact test file makes a difference, but the 2048 byte file I use (--blocksize=1024) is attached.

[2048.zip](https://github.com/duplicati/duplicati/files/12796989/2048.zip)
